File: src/main/java/com/springdebugger/convention/ConventionCheck.java
package com.springdebugger.convention;

import com.intellij.psi.PsiFile;

import java.util.List;

/**
 * One convention check implementation, selected by {@link ConventionRule#getCheckType()}. Each
 * implementation is a pure function of the file plus the rule's params: it reads the file's PSI and
 * returns the violations, with no IO and no cross-file resolution. That purity is what keeps the
 * checks deterministic and testable (see CONVENTIONS-PLAN.md section 4.3).
 */
public interface ConventionCheck {

    /** The checkType string this implementation handles, matching the catalog's `checkType` field. */
    String checkType();

    /** Find every violation of the given rule in the given file. */
    List<Violation> check(PsiFile file, ConventionRule rule);
}

File: vscode-extension/src/convention/convention-check.ts
// Port of com.springdebugger.convention.ConventionCheck. Each implementation is a pure function of
// the file text plus the rule's params: no IO, no cross-file resolution, deterministic and testable.
import { ConventionRule } from './convention-rule';
import { Violation } from './violation';

export interface ConventionCheck {
  /** The checkType string this implementation handles, matching the catalog's `checkType` field. */
  checkType(): string;

  /** Find every violation of the given rule in the given file text. */
  check(text: string, fileName: string, rule: ConventionRule): Violation[];
}
